cmake: make DaemonPlatform, Architecture, Compiler, etc. autonomous and reusable as Yokai framework

illwieckz opened the pull request:
Some revamp of the `DaemonPlatform`/`DaemonArchitecture`/`DaemonCompiler`/`DaemonBuildInfo` helper code for CMake.

I want this framework to be reusable for other projects, the first user outside of the engine may be the NaCl loader itself, and I may use it for NetRadiant as well at some point. So I made the code fully autonomous and contained in a single folder (so it can be copied and kept it sync easily).

It happens that the endian stuff we did in the `DaemonCompiler` code using `Endian.h` was useless (it was not doing better than what the `qprocessordetection.h` code does anyway), so I removed it.

I also renamed `GAME_PIE` to `NACL_PIE` because it's an NaCl option, it was named `GAME_PIE` because we only build games as NaCl, but for a self-contained framework we better name the NaCl option `NACL`. For now this code sets `GAME_PIE` to the `NACL_PIE` value but once the game is modified to use `NACL_PIE` the code setting the `GAME_PIE` alias will be deleted.

---

_Edit:_ That framework now has its own name: _**Yokai**_.

Yokai is the name of that reusable generic CMake framework. CMake variables are prefixed with `YOKAI_`, C variables and definitions are prefixed with `YOKAI_` as well, and the folder for generated files is prefixed with `Yokai`.

The name is inspired from that [third race mod](https://unvanquished.net/surviving-the-third-race/) that was named Unvanquished, from which we already inherited the Dæmon name (Dæmon was the third team name along Humans and Aliens, the Yokai model was one of those Dæmons).

illwieckz committed:
cmake: make DaemonPlatform, Architecture, Compiler, etc. autonomous and reusable

slipher commented:
Why would we need a variable called NACL_PIE? There is no PIE for NaCl and never will be.

illwieckz committed:
cmake: extend DaemonPlatform/System to detect host system and target system

slipher commented on cmake/DaemonPlatform/Platform.cmake:
```diff
@@ -0,0 +1,40 @@
+# Collection of reusable CMake helpers written for the Dæmon engine and related
+# projects.
+################################################################################
+
```
Why do we need this file? Are some of the files dependent on each other and so have to be included in a specific order?

illwieckz replied:
You'll end up loading them anyway.

The only thing that seems to be interesting to do is to provide an option to not generate the source files with the various discovered things, for projects not caring about reporting the build environment to the user. For example we don't need such reporting in NaCl.

illwieckz replied:
With the latest version, if you only include `Detection.cmake` without having included `SourceGenerator.cmake` first (or the `Platform.cmake` entry point which is meant to include everything supported), then no source is generated when detecting the build environment.

illwieckz committed:
cmake: prefix ARCH and NACL_ARCH with DAEMON

illwieckz committed:
cmake: remove unsupported platform thing, they're now considered unknown

We already renamed “unsupported” as “unknown” so it was just by luck
we didn't stop the build (what we wanted when changing that).

illwieckz committed:
cmake: use compiler detection instead of “MSVC” and “NOT MSVC AND WIN32”

illwieckz committed:
cmake: mutualize the detection framework

illwieckz committed:
cmake: do not write buildinfo files when SourceGenerator isn't included

illwieckz commented:
There is now a unique `DaemonPlatform/Detection/detect` script that can be used to test the compilation-based detection outside of CMake:

```
$ ./detect clang
DAEMON_COMPILER_GCC_COMPATIBILITY=ON
DAEMON_COMPILER_GCC_VERSION=4.2.1
DAEMON_COMPILER_Clang_COMPATIBILITY=ON
DAEMON_COMPILER_Clang_VERSION=18.1.3
DAEMON_COMPILER_Clang_VERSION_STRING=18.1.3 (1ubuntu1)
DAEMON_COMPILER_generic_VERSION_STRING=Ubuntu Clang 18.1.3 (1ubuntu1)
DAEMON_COMPILER_NAME=Clang
DAEMON_SYSTEM_Linux_COMPATIBILITY=ON
DAEMON_SYSTEM_Unix_COMPATIBILITY=ON
DAEMON_SYSTEM_NAME=Linux
DAEMON_ARCH_NAME=amd64
```

```
$ ./detect i686-w64-mingw32-gcc
DAEMON_COMPILER_GCC_COMPATIBILITY=ON
DAEMON_COMPILER_GCC_VERSION=13.0.0
DAEMON_COMPILER_generic_VERSION_STRING=13-posix
DAEMON_COMPILER_NAME=MinGW
DAEMON_SYSTEM_NAME=Windows
DAEMON_ARCH_NAME=i686
```

```
$ ./detect arm-nacl-clang
DAEMON_COMPILER_GCC_COMPATIBILITY=ON
DAEMON_COMPILER_GCC_VERSION=4.2.1
DAEMON_COMPILER_Clang_COMPATIBILITY=ON
DAEMON_COMPILER_Clang_VERSION=21.0.0
DAEMON_COMPILER_Clang_VERSION_STRING=21.0.0git (https://chromium.googlesource.com/native_client/nacl-llvm-project-v10.git 9c7f0369cfdd591e580c5ccfc1f00fedee58029f)
DAEMON_COMPILER_generic_VERSION_STRING=Clang 21.0.0git (https://chromium.googlesource.com/native_client/nacl-llvm-project-v10.git 9c7f0369cfdd591e580c5ccfc1f00fedee58029f)
DAEMON_COMPILER_NAME=Saigo
DAEMON_SYSTEM_Unix_COMPATIBILITY=ON
DAEMON_SYSTEM_NAME=NaCl
DAEMON_ARCH_NAME=nacl
```

illwieckz committed:
cmake: report native NaCl nexe architecture and “pnacl” for pexe

illwieckz committed:
cmake: generate printable DAEMON_SYSTEM_STRING

illwieckz committed:
cmake: unify the wording of Platform detected names

illwieckz commented:
As a side quest, I'm questioning the prefix. I wanted the variables produced by this framework to have a single prefix to 1. make it obvious it's generated by the framework, 2. have them share a namespace, 3. have them explicitly different with stock `CMAKE_` prefixed variables. Now, the `DAEMON_` prefix may not be the best, as we have genuine need for `DAEMON_` prefixed variables that don't need to be generated by this framework.

The thing is that I want this framework reusable. I plan to use it for the NaCl loader, I want to use it in NetRadiant, I may use it in XQF if the need arise. In those context, another name would be fine, though `DAEMON_` isn't bad because anyway it's the framework from Dæmon, but then it may look better if that's more obvious it's the platform than the engine itself. For example `DPLATFORM_` would also work.

illwieckz commented:
@slipher I would like to merge this so I can move on. Also I need to rebase the NaCl loader cmake code above it.

In the future I may find a more generic prefix than `DAEMON_` but for now it's good enough. If you think `DPLATFORM_` is fine or you have a better suggestion, I may do it. I still think it would be good if things from that platform would have a prefix different to the engine-specific stuff prefixed with `DAEMON_`. Actually we may go the bold way with a specific name (like Sloth, Urcheon…) for example to reuse [some name historically associated with the Unvanquished project](https://unvanquished.net/surviving-the-third-race/): `YOKAI_`.

illwieckz committed:
cmake: rename DaemonPlatform to Yokai, use custom variable slug

illwieckz committed:
cmake: rewrite comments

illwieckz committed:
yokai: drop custom variable slug, unify symbol naming

illwieckz commented:
Actually I just did that. This CMake framework was written for the Dæmon engine first but isn't specific to the engine and is meant to be reusable for any CMake-based project, like the NaCl loader, NetRadiant, etc.

So I gave it a name: _**Yokai**_. Yokai is the name of that reusable generic CMake framework. CMake variables are prefixed with `YOKAI_`, C definitions are prefixed with `YOKAI_` as well, and folders for generated files are prefixed with `Yokai`.

The name is inspired from that third race mod that was named Unvanquished, from which we already inherited the Dæmon name (Dæmon was the third team name along Humans and Aliens, the Yokai model was one of those Dæmons).

illwieckz commented:
One commit renames the framework to Yokai and but keeps a variable substitution mechanism to not break the existing `DAEMON_` prefixed things in C/C++ code and CMake code outside of Yokai, then the next commit removes the substitutions and fully embraces the Yokai namespace in the C/C++ and CMake code all the way down.

slipher commented:
> A good example of usage of them is to avoid the ambiguity of `WIN32`.
> 
> Instead of `WIN32` to detect the target we can use `DAEMON_TARGET_SYSTEM_Windows` and instead of `MSVC` and `NOT MSVC AND WIN32` to detect the compiler we can use `DAEMON_CXX_COMPILER_MSVC` and `DAEMON_CXX_COMPILER_MinGW`.

Please no. The built-in `WIN32` works fine. It is obnoxious to replace a standard language feature with a custom implementation. Anyone who was worked with CMake in other projects will not be able to use their knowledge. Anyone working on the project, familiar with CMake or not, will be inconvenienced by the use of a homemade version which does not appear in search engines, forcing them to read a bunch of crap to figure out what it is.

slipher commented on cmake/Yokai/SourceGenerator.cmake:
```diff
@@ -1,5 +1,5 @@
-# Daemon BSD Source Code
-# Copyright (c) 2025, Daemon Developers
+# YOKAI BSD Source Code
+# Copyright (c) 2025, YOKAI Developers
```
It's annoying to proliferate different author names because then we would have to update the author lists whenever moving things between files.

illwieckz replied:
Oh lol yes this one was just my `sed` command being not specific enough, I'll revert that. 😆️

illwieckz replied:
Fixed.

slipher commented on srclibs.cmake:
```diff
@@ -179,7 +179,7 @@ if (WIN32)
         ${LIB_DIR}/breakpad/src/client/windows/handler/exception_handler.h
     )
-elseif (LINUX)
+elseif (YOKAI_SYSTEM_Linux_COMPATIBILITY)
```
Does Breakpad Linux code seriously work on some non-Linux OS? Would be surprising since debugger APIs are very OS-specific. But if you want to try it I have a testing script: https://github.com/DaemonEngine/Daemon/pull/1785

illwieckz replied:
I totally forgot if I tested that on FreeBSD or not. Extending support isn't required for the framework to be merged so I returned to `SYSTEM_Linux`.

slipher commented on CMakeLists.txt:
```diff
@@ -534,7 +533,7 @@ else()
     # Look for OpenGL here before we potentially switch to looking for static libs.
     if (BUILD_CLIENT)
-        if (LINUX OR FREEBSD)
+        if (YOKAI_SYSTEM_Unix_COMPATIBILITY)
```
I assume "Unix compatibility" would include Mac? Then this is wrong because this GLVND stuff should not be set on Mac.

illwieckz replied:
Right, it doesn't intend to include macOS. I'm looking for a name for Linux-like systems, the ones with XDG desktops, Mesa drivers…

slipher replied:
What did you mean by having two different variables called Linux_COMPATIBILITY and Unix_COMPATIBILITY then?

Anyway my reaction is that the whole thing is a bad idea. Replacing `WIN32`, `LINUX`, `MSVC` with hand-rolled alternatives just makes the project hostile to contributors.

illwieckz replied:
Maybe I can just name it Linux compatibility, for things that are exactly the same as Linux, even if not Linux.

illwieckz replied:
Now named `SYSTEM_Linux_COMPATIBILITY`.

slipher commented on cmake/DaemonFlags.cmake:
```diff
@@ -604,15 +604,15 @@ endif()
 # Target options.
-if (MSVC)
-    if (ARCH STREQUAL "i686")
+if (YOKAI_CXX_COMPILER_MSVC)
```
Replacing MSVC conditionals with this breaks the `clang-cl` compiler which is both Clang and MSVC. It is correctly detected by CMake:
```
-- The C compiler identification is Clang 19.1.5 with MSVC-like command-line
-- The CXX compiler identification is Clang 19.1.5 with MSVC-like command-line
```

clang-cl works on master but this PR breaks it by feeding it GCC-specific options.

Was there anything wrong with CMake's compiler detection that you saw?

illwieckz replied:
This is a perfect example of why I wrote this code, to avoid such CMake ambiguity.

When rewriting that, I made the wrong assumption that “MSVC” meant “MSVC compiler” and not “MSVC compiler compatibility”. To avoid this mistake of mine I would have to read CMake docs. But to feel the need to read CMake docs, it was required from to to have in mind the knowledge that there exists non-MSVC compilers compatible with MSVC, information I totally forgot when I rewrote this.

With this framework which is meant to remove ambiguity from CMake, someone reading `COMPILER_MSVC_COMPATIBILITY` doesn't need to read any doc to know what's happening: it's for any compiler that behaves like MSVC.

So the same way I already implemented `COMPILER_GCC_COMPATIBILITY` and `COMPILER_Clang_COMPATIBILITY`, I now implemented `COMPILER_MSVC_COMPATIBILITY` and I used it where ambiguous and misleading `MSVC` was used.

And `COMPILER_MSVC` can be used to target non-Clang MSVC if needed.

slipher replied:
Because no docs at all is so much better than reading CMake docs!

slipher replied:
> When rewriting that, I made the wrong assumption that “MSVC” meant “MSVC compiler” and not “MSVC compiler compatibility”. To avoid this mistake of mine I would have to read CMake docs. But to feel the need to read CMake docs, it was required from to to have in mind the knowledge that there exists non-MSVC compilers compatible with MSVC, information I totally forgot when I rewrote this.
> 
> With this framework which is meant to remove ambiguity from CMake, someone reading `COMPILER_MSVC_COMPATIBILITY` doesn't need to read any doc to know what's happening: it's for any compiler that behaves like MSVC.

Seems backward to me. 99% of the time what you want is the MSVC-compatible options/ABI thing. So the person who is blissfully unaware of `clang-cl` will write `if (MSVC)` and get it right, even if by accident. While in the proposed framework they will have multiple options like YOKA_MSVC_COMPATIBILITY or YOKAI_COMPILE_MSVC and they can just as easily get it wrong.

slipher commented:
Besides clang-cl being broken, regular old MSVC also doesn't work. 

``` NaCl VM is not supported on this platform. Set vm.cgame.type and vm.sgame.type to 3 (native DLL) and use devmap instead of map.```

illwieckz commented:
> Besides clang-cl being broken, regular old MSVC also doesn't work.
> 
> ` NaCl VM is not supported on this platform. Set vm.cgame.type and vm.sgame.type to 3 (native DLL) and use devmap instead of map.`

This wasn't MSVC-specific, it was a mistake that happened when rebasing over the box64 code.
By merging this sooner than later, we would avoid such kind of rebase mistakes to pile up.

illwieckz commented:
> > A good example of usage of them is to avoid the ambiguity of `WIN32`.
> > Instead of `WIN32` to detect the target we can use `DAEMON_TARGET_SYSTEM_Windows` and instead of `MSVC` and `NOT MSVC AND WIN32` to detect the compiler we can use `DAEMON_CXX_COMPILER_MSVC` and `DAEMON_CXX_COMPILER_MinGW`.
> 
> Please no. The built-in `WIN32` works fine. It is obnoxious to replace a standard language feature with a custom implementation. Anyone who was worked with CMake in other projects will not be able to use their knowledge. Anyone working on the project, familiar with CMake or not, will be inconvenienced by the use of a homemade version which does not appear in search engines, forcing them to read a bunch of crap to figure out what it is.

As said [above](https://github.com/DaemonEngine/Daemon/pull/1641#discussion_r3362946597), CMake is ambiguous and not only requires to read docs, but is ambiguous in a way it even doesn't give the incentive to read docs.

Things like `COMPILER_MSVC_Compatibility` or `SYSTEM_Windows` are very much less ambiguous and don't need to read any doc to be understood. That's the whole purpose.

If we want to remove even more ambiguity we may rename to `TARGET_SYSTEM_Windows`. We currently have `HOST_SYSTEM_Windows` and `SYSTEM_Windows`, but the  `SYSTEM_Windows` keyword being for the target isn't context-free for the reader.

illwieckz commented:
>  Anyone who was worked with CMake in other projects will not be able to use their knowledge.

Someone writing new code using `WIN32` or `MSVC` will still get his CMake code working, he will be able to use his knowledge, but he will just introduce more error-prone code like we did before.

We would be able to encourage such contributor to use our unambiguous symbols when reviewing, or just rewrite his code later to make it more robust.

Writers don't lose their knowledge, and readers would even require _less_ knowledge.

slipher commented on cmake/Yokai/Architecture.cmake:
```diff
@@ -0,0 +1,128 @@
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
```
more license junk

illwieckz replied:
Is the junk being the `<COPYRIGHT HOLDER>` unexpanded template?

This template is used unexpanded in:

- CMakeLists.txt
- cmake/DaemonBuildTypeGeneratorExpression.cmake
- cmake/FindWasmtime.cmake
- cmake/DaemonNacl.cmake
- cmake/toolchain-pnacl.cmake
- cmake/DaemonGame.cmake
- cmake/DaemonFlags.cmake
- cmake/DaemonCBSE.cmake

I can fix it in Yokai's files, but that is something generic to our CMake code.

slipher replied:
> Is the junk being the `<COPYRIGHT HOLDER>` unexpanded template?

right

illwieckz replied:
I expanded the header template for all our CMake files.

slipher commented on cmake/Yokai/Architecture.cmake:
```diff
@@ -0,0 +1,128 @@
+		if ("${arch_name}" IN_LIST armhf_usage)
+		# Load 32-bit armhf nexe on 64-bit arm64 engine on Linux with multiarch.
+		# The nexe is system agnostic so there should be no difference with armel.
+		set(nacl_arch "armhf")
```
I imagine this NaCl and box64 stuff shouldn't be in here as it's Daemon-specific.

illwieckz replied:
I was thinking about only running this NaCl setup code when some variable is set to enable it.

In the same way the Yokai framework can already be used without using the source generation code, it would be usable without caring about NaCl.

illwieckz replied:
This code is likely not needed when called from `DaemonGame.cmake` to begin with.

illwieckz replied:
This NaCl code is now only used when `YOKAI_NACL_HOST` is enabled.

The game code is the first Yokai consumer not using `YOKAI_NACL_HOST`, and the Dæmon engine is the first Yokai consumer using `YOKAI_NACL_HOST`.

slipher replied:
I don't get why you don't just move that stuff somewhere outside of the "Yokai" directory.

illwieckz replied:
Engine:

```
$ cat YokaiGeneratedSource/YokaiBuildInfo/Engine.cpp 
// Automatically generated, do not modify!
const char* YOKAI_SYSTEM_STRING="Linux";
const char* YOKAI_ARCH_STRING="amd64";
const char* YOKAI_NACL_ARCH_STRING="amd64";
const char* YOKAI_C_COMPILER_STRING="GCC_13.3.0:cc";
const char* YOKAI_CXX_COMPILER_STRING="GCC_13.3.0:c++";
```

Dll game:

```
$ cat YokaiGeneratedSource/YokaiBuildInfo/Game.cpp
// Automatically generated, do not modify!
const char* YOKAI_SYSTEM_STRING="Linux";
const char* YOKAI_ARCH_STRING="amd64";
const char* YOKAI_C_COMPILER_STRING="GCC_13.3.0:gcc";
const char* YOKAI_CXX_COMPILER_STRING="GCC_13.3.0:g++";
```

Nexe game:

```
$ cat nacl-vms/YokaiGeneratedSource/YokaiBuildInfo/Game.cpp 
// Automatically generated, do not modify!
const char* YOKAI_SYSTEM_STRING="NaCl";
const char* YOKAI_ARCH_STRING="pnacl";
const char* YOKAI_C_COMPILER_STRING="PNaCl_4.2.1/Clang_3.6.0:pnacl-clang";
const char* YOKAI_CXX_COMPILER_STRING="PNaCl_4.2.1/Clang_3.6.0:pnacl-clang++";
```

slipher commented on src/common/Common.h:
```diff
@@ -31,6 +31,14 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#if !defined(DAEMON_ARCH_STRING)
+#define DAEMON_ARCH_STRING YOKAI_ARCH_STRING
+#endif
+#if !defined(YOKAI_CXX_COMPILER_STRING)
```
this doesn't make sense

illwieckz replied:
Fixed.

slipher commented:
> If we want to remove even more ambiguity we may rename to `TARGET_SYSTEM_Windows`. We currently have `HOST_SYSTEM_Windows` and `SYSTEM_Windows`, but the `SYSTEM_Windows` keyword being for the target isn't context-free for the reader.

At least that would *actually* be less ambiguous than what we have. The `SYSTEM_` word by itself does not provide any information.

slipher commented on CMakeLists.txt:
```diff
@@ -463,23 +462,23 @@ endif()
 # Native client
 include(DaemonNacl)
-if (NACL)
+if (YOKAI_SYSTEM_NaCl)
```
Requesting the system names to be all caps. Trying to spell this mixed case gibberish like SYSTEM_NaCl and SYTEM_macOS seems painful.

illwieckz replied:
Well, the “NaCl” and “macOS” strings are also stored as `const char* YOKAI_SYSTEM_STRING` which is meant to be printed and then be pretty, but I can add code to upper case them when making a CMake variable or a C macro out of them.

illwieckz replied:
Now that a lot of code has been made reusable and genericized accross System, Architecture and Compiler, it's probably even easier to do so.

illwieckz replied:
Done. Now both CMake variables and C macros are fully capitalized.

illwieckz committed:
yokai: only run NaCl code when YOKAI_NACL_HOST is enabled

illwieckz committed:
yokai: only process true MSVC as MSVC

illwieckz committed:
yokai: generate capitalized CMake variable names and C macro names

illwieckz committed:
cmake: generate capitalized C macro name

illwieckz committed:
yokai: expand the header template

illwieckz committed:
cmake: expand the header template

illwieckz committed:
yokai: header date

illwieckz commented:
> > If we want to remove even more ambiguity we may rename to `TARGET_SYSTEM_Windows`. We currently have `HOST_SYSTEM_Windows` and `SYSTEM_Windows`, but the `SYSTEM_Windows` keyword being for the target isn't context-free for the reader.
> 
> At least that would _actually_ be less ambiguous than what we have. The `SYSTEM_` word by itself does not provide any information.

I think the reason why I haven't wrote `TARGET_SYSTEM` but `SYSTEM` is because the symbol name is currently the same in CMake and in C/C++ file, and in C/C++ file it's always a target system…

illwieckz committed:
yokai: use TARGET_SYSTEM

illwieckz committed:
yokai: use TARGET_ARCH

illwieckz commented:
The target symbols are now `YOKAI_TARGET_SYSTEM` and `YOKAI_TARGET_ARCH`, they are just `YOKAI_SYSTEM` and `YOKAI_ARCH` in C/C++ (not `TARGET`).

illwieckz committed:
yokai: some prefix rework

slipher commented on cmake/Yokai/All.cmake:
```diff
@@ -1,5 +1,5 @@
 # Daemon BSD Source Code
-# Copyright (c) 2013-2016, Daemon Developers
+# Copyright (c) 2025-2006, Daemon Developers
```
typo

illwieckz replied:
Fixed.

slipher commented on cmake/Yokai/System.cmake:
```diff
@@ -0,0 +1,131 @@
+# > if (YOKAI_HOST_SYSTEM_LINUX_COMPATIBILITY)
+# > if (YOKAI_TARGET_SYSTEM_LINUX_COMPATIBILITY)
+foreach(name LINUX;FREEBSD)
+	foreach(slug HOST;HOST_SYSTEM)
```
I think you wanted TARGET to be one of the things in this line?

Just write it without loops, it would be shortesr that way anyway.

illwieckz replied:
Now doing `HOST_SYSTEM;TARGET_SYSTEM`. Indeed `HOST` was never correct.

illwieckz replied:
In fact, we should not name that `YOKAI_TARGET_SYSTEM_LINUX_COMPATIBILITY`, `YOKAI_TARGET_SYSTEM_LINUX_COMPATIBILITY` was set twice, first as a marker of `__linux__` defined in preprocessor (matching both Linux and Android), and this other one meant to match Linux-like desktops (operating system environment). So this one is now named `YOKAI_TARGET_SYSTEM_XDG_COMPATIBILITY`.

So now there is:

- `YOKAI_TARGET_SYSTEM_LINUX_COMPATIBILITY` is for when there is Linux API (`__linux__`), being a Linux desktop or not.
- `YOKAI_TARGET_SYSTEM_XDG_COMPATIBILITY` is for systems behaving like a Linux desktop (FHS, XDG desktop, GLVND…)

I named it XDG because I wanted some standard, not an operating system. XDG (X desktop group) is about desktops, but a system with an XDG desktop (GNOME, KDE…) very likely also do FHS, GLVND, uses `.a` for static library, things like that. So I thought that “XDG compatibility” is very close to what it is meant.

slipher commented on cmake/DaemonFlags.cmake:
```diff
@@ -263,7 +264,7 @@ if (NOT NACL AND BUILD_CLIENT)
 endif()
 
-if (MSVC)
+if (YOKAI_CXX_COMPILER_MSVC_COMPATIBILITY)
```
`/MP` doesn't do anything and prints a warning in the Clang clone so this would be a good opportunity to use a really real MSVC variable.

illwieckz replied:
Now only doing `/MP` when MSVC is true MSVC (using `YOKAI_CXX_COMPILER_MSVC` which is only set when `CMAKE_CXX_COMPILER_ID STREQUAL "MSVC"`).

slipher commented on srclibs.cmake:
```diff
@@ -179,7 +179,9 @@ if (WIN32)
     )
-elseif (LINUX)
+# We don't know if this works outside of Linux,
+# so we don't use YOKAI_TARGET_SYSTEM_LINUX_COMPATIBILITY.
```
It definitely won't work, see https://forums.freebsd.org/threads/porting-google-breakpad.18039/

illwieckz replied:
Comment removed.

illwieckz replied:
In fact it was indeed meant to be `LINUX_COMPATIBILITY`, so it can be used on Android even if not a Linux Desktop.

It's just that I defined twice `LINUX_COMPATIBILITY`, and the second one with another meaning:

- https://github.com/DaemonEngine/Daemon/pull/1641#discussion_r3377513626

Now I use `XDG_COMPATIBILITY` when the desktop is linux-like, while `LINUX_COMPATIBILITY` means the Linux API is actually there.

And for completeness, `LINUX` means Linux desktop, not Android (that one of the ambiguity I also wanted to remove).

slipher commented on CMakeLists.txt:
```diff
@@ -733,7 +733,7 @@ if (USE_BREAKPAD)
             srclibs-breakpad-crash_generation_client srclibs-breakpad-common)
 
-    elseif (LINUX)
+    elseif (YOKAI_TARGET_SYSTEM_LINUX_COMPATIBILITY)
```
should be exactly Linux

illwieckz replied:
Done.

illwieckz replied:
Same as said in other comments, there was a name collision.

In fact I was targeting the Linux API (`__linux__`) there, but the `(FREEBSD OR LINUX)` variable was mistakenly named `LINUX_COMPATIBILITY` too, bringing confusion.

slipher replied:
Why does it still say COMPATIBILITY? What would be the difference between that and YOKAI_TARGET_SYSTEM_LINUX now?

illwieckz replied:
`YOKAI_TARGET_SYSTEM_LINUX` reports Linux desktop and is meant to be consistent with `YOKAI_SYSTEM_STRING` reporting `"Linux"` in C++ code.

While `YOKAI_TARGET_SYSTEM_ANDROID` reports Android and is meant to be consistent with `YOKAI_SYSTEM_STRING` reporting `"Android"` in C++ code.

The `YOKAI_TARGET_SYSTEM_LINUX_COMPATIBILITY` is the equivalent of `__linux__`, matching both Linux and Android.

When Android is matched, Android is reported, not Linux, but Linux compatibility is reported.

This is similar to the compiler thing where both Clang, ArmClang, AppleClang, AOCC and ICX reports Clang compatibility, but we can still match upstream Clang among its forks, or one of those forks.

It makes even more sense for Linux because between things like Linux and Android the userland is way more different than between two Clang downstreams.

slipher commented:
I should mention that besides the Clang that pretends to be MSVC, there is also [Clang that pretends to be MinGW](https://wiki.unvanquished.net/wiki/Compiling_the_source#MSYS2_Clang). I don't have it installed currently so I can't say how the compiler is detected.

slipher commented:
I still don't understand the claimed problems with making each cmake file self-contained (by including dependencies using the `include_guard` thing), or doing NaCl stuff outside of Yokai. `daemon_generate_buildinfo` is called in Daemon's CMakeLists.txt, and it's still that way after the refactoring. That's a good thing: clients can decide whether they want a build info to be generated, and add their own definitions to it. So making the buildinfo cmake file a dependency of any other file is no problem because it doesn't contribute anything to the build unless `generate_buildinfo` is called. And Daemon can add its own build infos respecting NaCl after including the Yokai scripts.

(Also I think it's bad how build info is done with gamelogic currently, though that can go in another PR. `DaemonGame.cmake` currently generates the build info when it's included and then it's hackily included from `q_shared.h` with ifdefs. Instead of that, the gamelogic should call `daemon_generate_buildinfo` itself (if it even wants to) and include it from its own files: that way it would have the chance to add its own definitions.)

illwieckz committed:
cmake: don't use /MP on clang-cl

illwieckz committed:
cmake: move NaCl host code out of Yokai

illwieckz commented:
The NaCl architecture stuff is now removed from Yokai and stored in `cmake/DaemonNaclHost.cmake`, it is only called by the engine CMake code. Note that I will probably deeply rewrite this code later, but that's outside of the scope if this PR.

illwieckz commented:
> (Also I think it's bad how build info is done with gamelogic currently, though that can go in another PR. `DaemonGame.cmake` currently generates the build info when it's included and then it's hackily included from `q_shared.h` with ifdefs. Instead of that, the gamelogic should call `daemon_generate_buildinfo` itself (if it even wants to) and include it from its own files: that way it would have the chance to add its own definitions.)

Right now Dæmon's `CMakeLists.txt` does:

```cmake
yokai_write_buildinfo("DaemonEngine")
```

And game's `DaemonGame.cmake` does:

```cmake
yokai_write_buildinfo("DaemonGame")
```

The buildinfo is never mixed between engine and game.

illwieckz commented:
> The buildinfo is never mixed between engine and game.

Also the list of things in engine and game buildinfo aren't the same anymore, for example the game buildinfo doesn't have the `DAEMON_NACL_ARCH_STRING` constant that is engine-only.

Also it's not possible for the game to use engine's buildinfo because the nexe build is done in a subfolder.

The name passed to `yokai_write_buildinfo` is the name of the subfoler the buildinfo data is written to, to make sure the game doesn't reuse engine buildinfo by mistake when built as dll/exe (and sharing the build folder with the engine).

slipher commented:
> > (Also I think it's bad how build info is done with gamelogic currently, though that can go in another PR. `DaemonGame.cmake` currently generates the build info when it's included and then it's hackily included from `q_shared.h` with ifdefs. Instead of that, the gamelogic should call `daemon_generate_buildinfo` itself (if it even wants to) and include it from its own files: that way it would have the chance to add its own definitions.)
> 
> Right now Dæmon's `CMakeLists.txt` does:
> 
> ```cmake
> yokai_write_buildinfo("DaemonEngine")
> ```
> 
> And game's `DaemonGame.cmake` does:
> 
> ```cmake
> yokai_write_buildinfo("DaemonGame")
> ```
> 
> The buildinfo is never mixed between engine and game.

I'm saying the write_buildinfo should be in Unvanquished CMakeLists.txt, not in DaemonGame.cmake. I'm hoping this could fix https://github.com/DaemonEngine/Daemon/issues/1599.

slipher commented on cmake/Yokai/Detection/System.c:
```diff
@@ -0,0 +1,72 @@
+#endif
+
+#if defined(__unix__) || defined(__unix) || defined(unix)
+	#pragma message(REPORT_COMPATIBILITY("Unix"))
```
I suggest removing this as it has unexpected semantics: Mac OS is a Unix, but does not define these. Anyhow the good old idiom for Unix-like is `NOT WIN32` :)

slipher commented on cmake/Yokai/System.cmake:
```diff
@@ -0,0 +1,131 @@
+# implementing standards like FHS, XDG, GLVND…
+# Makes possible to do that in CMake code:
+# > if (YOKAI_HOST_SYSTEM_XDG_COMPATIBILITY)
+# > if (YOKAI_SYSTEM_XDG_COMPATIBILITY)
```
Instead of XDG_COMPATIBILITY I prefer the GNU_COMPATIBILITY name you used in some other places. This makes sense by reflecting the "GNU/Linux" userland/kernel distinction. So we have Linux compatibility include Android's Linux-based kernel, and GNU compatibility including the FreeBSD using a similar userland to the Linux desktops.

slipher replied:
Well it's not that great since GNU compatibility could also be confused for the "GNU compiler id" that CMake identifier for GCC. But at least it relates to the "GNU/Linux" meme. For me XDG doesn't suggest the concept you want to get at at all.

illwieckz replied:
In some iterations I wrote GNU too, then switched to XDG. The thing is that GNU means GNU Hurd in System context.

One may say that GNU Linux is probably similar to GNU Hurd, not less then FreeBSD, but well…

The `SYSTEM_GNU_COMPATIBILITY` in some other places are leftovers remaining from that first iteration I failed to clean-up properly. I'll fix it.

illwieckz replied:
Fixed.

slipher commented on cmake/DaemonNaclHost.cmake:
```diff
@@ -0,0 +1,78 @@
+
+		set(box64_usage ppc64el riscv64)
+		if ("${arch_name}" IN_LIST box64_usage)
+			option(YOKAI_NACL_BOX64_EMULATION "Use Box64 to emulate x86_64 NaCl loader on unsupported platforms" ON)
```
can go back to its original name

illwieckz replied:
Especially since I renamed it in C++ code already. 😆️

Now fixed.

illwieckz committed:
yokai: unify a bit the detect test script output with CMake

illwieckz commented:
> I should mention that besides the Clang that pretends to be MSVC, there is also [Clang that pretends to be MinGW](https://wiki.unvanquished.net/wiki/Compiling_the_source#MSYS2_Clang). I don't have it installed currently so I can't say how the compiler is detected.

Right now this does:

```
#elif defined(__MINGW64__) || defined(__MINGW32__)
	#pragma message(REPORT_NAME("MinGW"))
#elif defined(__clang__)
	#pragma message(REPORT_NAME("Clang"))
```

And this doesn't do any:

```
#pragma message(REPORT_COMPATIBILITY("MinGW"))
```

So for now I expect it to just be reported as `COMPILER_MINGW` with `COMPILER_CLANG_COMPATIBILITY`.

If needed we can make it be reported as `COMPILER_CLANG` with `COMPILER_MINGW_COMPATIBILITY` instead, or something like that.

One can use the `Yokai/Detection/detect` script for a quick preview of what CMake is expected to provide:

```
$ ./cmake/Yokai/Detection/detect clang
YOKAI_C_COMPILER_GCC_COMPATIBILITY=ON
YOKAI_C_COMPILER_GCC_VERSION=4.2.1
YOKAI_C_COMPILER_CLANG_COMPATIBILITY=ON
YOKAI_C_COMPILER_CLANG_VERSION=18.1.3
YOKAI_C_COMPILER_CLANG_VERSION_STRING=18.1.3 (1)
YOKAI_C_COMPILER_GENERIC_VERSION_STRING=Ubuntu Clang 18.1.3 (1)
YOKAI_C_COMPILER_NAME=Clang
YOKAI_TARGET_SYSTEM_LINUX_COMPATIBILITY=ON
YOKAI_TARGET_SYSTEM_UNIX_COMPATIBILITY=ON
YOKAI_TARGET_SYSTEM_NAME=Linux
YOKAI_TARGET_ARCH_NAME=amd64

$ ./cmake/Yokai/Detection/detect aarch64-linux-android29-clang
YOKAI_C_COMPILER_GCC_COMPATIBILITY=ON
YOKAI_C_COMPILER_GCC_VERSION=4.2.1
YOKAI_C_COMPILER_CLANG_COMPATIBILITY=ON
YOKAI_C_COMPILER_CLANG_VERSION=9.0.8
YOKAI_C_COMPILER_CLANG_VERSION_STRING=9.0.8 (https://android.googlesource.com/toolchain/llvm-project 207d7abc1a2abf3ef8d4301736d6a7ebc224a290)
YOKAI_C_COMPILER_GENERIC_VERSION_STRING=4.2.1 Compatible Android (5900059 based on r365631c) Clang 9.0.8 (https://android.googlesource.com/toolchain/llvm-project 207d7abc1a2abf3ef8d4301736d6a7ebc224a290)
YOKAI_C_COMPILER_NAME=Clang
YOKAI_TARGET_SYSTEM_LINUX_COMPATIBILITY=ON
YOKAI_TARGET_SYSTEM_UNIX_COMPATIBILITY=ON
YOKAI_TARGET_SYSTEM_NAME=Android
YOKAI_TARGET_ARCH_NAME=arm64

$ ./cmake/Yokai/Detection/detect i686-w64-mingw32-gcc
YOKAI_C_COMPILER_GCC_COMPATIBILITY=ON
YOKAI_C_COMPILER_GCC_VERSION=13.0.0
YOKAI_C_COMPILER_GENERIC_VERSION_STRING=13-posix
YOKAI_C_COMPILER_NAME=MinGW
YOKAI_TARGET_SYSTEM_NAME=Windows
YOKAI_TARGET_ARCH_NAME=i686
```

slipher commented on cmake/Yokai/Detection.cmake:
```diff
@@ -0,0 +1,110 @@
+
+# Make sure to always call this macro from within a function, not in the global scope.
+# As a macro it produces a lot of variables in the parent scope but it is meant to
+# only be called by functions so they should never pollute the globale scope.
```
"globale"

illwieckz replied:
Fixed.

slipher commented on cmake/Yokai/Compiler.cmake:
```diff
@@ -25,66 +25,21 @@
-				PARENT_SCOPE)
-		endforeach()
-	endif()
+	yokai_run_detection("${lang}" "COMPILER" "Compiler${${lang}_EXT}" "GCC;Clang;generic")
```
"generic" string seems to be unused

illwieckz replied:
It exports compiler's `__VERSION__` macro to `YOKAI_C_COMPILER_GENERIC_VERSION_STRING`.

This is used (named as temporary variable `compiler_generic_version_string` in `yokai_detect_compiler` function) to extract the PNaCl version string.

slipher commented on cmake/DaemonFlags.cmake:
```diff
@@ -382,7 +386,7 @@ else()
 	endif()
 
-	if (NACL AND USE_NACL_SAIGO AND SAIGO_ARCH STREQUAL "arm")
+	if (YOKAI_TARGET_SYSTEM_NACL AND USE_NACL_SAIGO AND YOKAI_NACL_ARCH_armhf)
```
missed capitalization spot?

illwieckz replied:
Fixed.

slipher commented on cmake/DaemonFlags.cmake:
```diff
@@ -625,25 +629,25 @@ elseif (NOT NACL)
+		if (YOKAI_CXX_COMPILER_ICC)
 			set(GCC_GENERIC_ARCH "pentium4")
-		elseif (DAEMON_CXX_COMPILER_Zig)
+		elseif (YOKAI_CXX_COMPILER_Zig)
```
missed capitalization spot?

illwieckz replied:
Fixed.

slipher commented:
Will these commits be squashed?

There's one outstanding comment about the UNIX_COMPATIBILITY variable having bad semantics.

illwieckz commented:
> Will these commits be squashed?

They are fine as they are right now. Every time a mistake has been found, I applied the fix to the earliest commit, fixing the rebase all the way up, in order for every commit to be correct. The only commit I didn't do that yet was the `cmake: fix a saigo armhf typo` and I now done it. Each commits are autonomous and evolving the stuff step by step.

> There's one outstanding comment about the UNIX_COMPATIBILITY variable having bad semantics.

The remaining generated `UNIX_COMPATIBILITY` variable just means `#if defined(__unix__) || defined(__unix) || defined(unix)`.

slipher commented:
> The remaining generated `UNIX_COMPATIBILITY` variable just means `#if defined(__unix__) || defined(__unix) || defined(unix)`.

And that's bad because Mac is a Unix platform but doesn't define those.

slipher commented:
This comment: https://github.com/DaemonEngine/Daemon/pull/1641#discussion_r3381095138

illwieckz commented:
> > The remaining generated `UNIX_COMPATIBILITY` variable just means `#if defined(__unix__) || defined(__unix) || defined(unix)`.
> 
> And that's bad because Mac is a Unix platform but doesn't define those.

Are you sure? I don't have an easily accessed mac next to me right now, but if you have one it's easy to check by doing:

```
echo '' | cc -dM -E -
```

If macOS doesn't define those then it explicitly refuses to be recognized as an Unix platform and we should respect that…

It borrows things from earlier MacOS, some BSD Unix stuff, GNU, NextSTEP and other things. Pre-installing a 20 years old bash (even not used as shell by default) doesn't make something an Unix. =)

slipher commented:
> Are you sure? I don't have an easily accessed mac next to me right now, but if you have one it's easy to check by doing:
> 
> ```
> echo '' | cc -dM -E -
> ```

I confirm that none of the Unix defines appear from this command.

> If macOS doesn't define those then it explicitly refuses to be recognized as an Unix platform and we should respect that…

https://www.opengroup.org/openbrand/certificates/1205p.pdf

Mac OS is universally recognized as a Unix-like system. You have said that the names are intended to be understood without reading anything, so it behooves us not to make one that directly contradicts its usual understanding.

illwieckz commented:
Well, then we may drop this `__unix__` macro detection entirely. We don't use it for Dæmon anyway, the NaCl loader doesn't use it as well. Some of our dependencies use it, like LZMA in Crunch, and some tests in Recast (but then that doesn't contribute to the binary). This `__unix__` macro seems to be useless and unusable.

illwieckz commented:
Now done.

I also fixed something: we're not supposed to use `NACL_ARCH_XXX` when `NOT TARGET_SYSTEM_NACL`, but `TARGET_ARCH_XXX` instead.

illwieckz commented:
I wanted to test with Saigo and I just did and it worked.

slipher commented:
I think the Breakpad Linux build is broken.

slipher commented:
Never mind, it's the combination of Breakpad and dummy app, which is already broken on master.

slipher commented:
LGTM

illwieckz commented:
@slipher about the `include_guard` thing, I haven't investigated it because it wasn't needed, but if someone thinks it can be used to improve things, that one can investigate this.
